### ISSUE TITLE
EU model

### ISSUE THREAD

**Issue by jorgtied** [Helsinki-NLP/mammoth#91]
Building a multilingual modular model for translating between the official EU languages
* initial model with basic architecture and English-centric data
* test runs with selected architectures with different amounts of parameter sharing and model sizes
* final EU model that can be used as starting point for language/task extension

**Comment by jorgtied** [comment]
#98 provides the relevant training data
first iteration architecture:
* language-specific decoder(3 layers)
* fully shared encoder (9 layers)
* English-centric only
* Auto-encodder task (different denoising losses)

Goal: initial model trained by July 15

**Comment by jorgtied** [comment]
Let's get back to training the EU model. As stated before, architecture would be
* fully-shared encoder (9 layers)
* language-specific decoders (3 layers)

Data and training tasks:
* 37 languages, English-centric in both directions (English-to-xx and xx-to-English)
* synthetic parallel data produced by translating nemotron-cc available from https://github.com/Helsinki-NLP/translate-fineweb
* add auto-encoder task to improve transfer learning

Dataset downloads are in plain text format, aligned with English (corresponding rows), for example, for English-Finnish, we have https://github.com/Helsinki-NLP/translate-fineweb/blob/main/maxidl/nemotron-cc-english-run1/translated/txt/fin/README.md

* English: https://object.pouta.csc.fi/OELLM-synthetic/maxidl/nemotron-cc-english-run1/translated/txt/eng/nemotron-cc-english-run1-train-00.txt.gz
* Finnish: https://object.pouta.csc.fi/OELLM-synthetic/maxidl/nemotron-cc-english-run1/translated/txt/fin/nemotron-cc-english-run1-train-00.txt.gz

There are 100 shards, roughly 1 billion tokens per shard. Starting with one shard across all languages should probably be more than enough for the first runs.

**Comment by amikael** [comment]
**Background**

We have intuitions and know-how on how the tasks should be distributed and what kind of model architectures usually work best.  Config-config is an attempt to formalize the GPU assignment problem in terms of an objective function that minimizes the communication overhead.  But the parameters of the objective function are guesses that may be far from the best fit with the empirical overheads.  The main part of the objective function models the effect of the distribution of simultaneous tasks, since loading or saving complete models will not be as costly as multiple epochs of training and the related gradient summing and parameter updates within sharing layers.

The temporal distribution of the tasks in a cluster may also affect the learning rate or translation quality when all tasks do not fit simultaneously on the GPUs.  But exploring this dimension of scaling requires several time slots for the tasks, which is too much for the first experiments.

**The aim of the experiments** 

"to find out how the temporary GPU assignment affects the training"

**Input variables**

- alternative ways to distribute the tasks to nodes:
> (1) [en-lang, lang-en, lang-lang]x2 = 6 tasks per node,
> (2) [en-lang1, lang1-lang1, en-lang2, lang2-lang2, ..] = 8 tasks per node, 
> (3) [en-lang1, en-lang2, ... ] + [lang1-en, lang2-en, ...] + [lang1-lang1, lang2-lang2,...]

The last task distribution is so odd that it is unlikely that it has never been mentioned in the literature.  However, from the scientific perspective, even basic results on the design choices can be cited later.
 
- growing number of languages:
> (1) one node (en-en + 2x [en-lang, lang-en, lang-lang] = 7 tasks)
> (2) two nodes (en-en + 5x [en-lang, lang-en, lang-lang] = 16 tasks)
> (3) four nodes (en-en + 10x [en-lang, lang-en, lang-lang] = 31 tasks)
> (4) eight nodes (en-en + 21x [en-lang, lang-en, lang-lang] = 64 tasks)

**output variables**

> - communication overhead
> - learning rate
> - quality scores
> - scalability in terms of the number of languages

**Methods:**

> - Mammoth (main branch), the described architecture, LUMI, up-to-date environment
> - one task per GPU, one time slot (shard) => temporally flat set-up
> - high enough debug level to get all information about inter-device communication ops and amounts
> - mining of the cost distribution from the log files
> - modelling the cost as a function of the type of the task distribution over the cluster.

**Comment by amikael** [comment]
I checked [https://cameronbarker.uk/posts/intro-to-hf-tokenizers/](https://cameronbarker.uk/posts/intro-to-hf-tokenizers/) to learn about HF tokenizers.
There are a few choices to be made:
- **Which model?**  BPE, Unigram, WordLevel and WordPiece?
- **Vocabulary size?**  "In Llama, 3 the size 128256  corresponds to 128000 tokens plus 256 special tokens." Chao's script uses 180000. Earlier Mammoth tutorials (https://helsinki-nlp.github.io/mammoth/prepare_data.html) build sentence piece tokenizers with 65k tokens.  For a bilingual tokenizer, 32000 or 64000 should be OK.  Chao offered a number of tokenizers, but these use 262144 tokens and do not cover all languages I am working on.
- **Normalization:** lowercasing? removing leading and trailing spaces?
- **Pretokenization:** I observed that the tokenizer can include punctuation to common words:  ['▁Hello,', '▁world!', '▁This', '▁is', '▁a', '▁test.'].  There are a lot of arbitrary combinations in the lexicon: "________": 10664 (ruler),  "as,": 12303, ".)\n": 12304 (punctuations), "▁website!": 37480 (punctuated words), "▁40,000": 37481 (numbers).   This does not seem preferrable. "In Llama 3, the pretokenizer is a sequence of two pretokenizers: Split and ByteLevel. Split is used to apply a regex pattern to the text. See https://regex101.com/r/m6gEfJ/1 for a breakdown of the regex pattern used in Llama 3. Loosely, the pattern splits before whitespace, punctuation and numbers."
- **Multi-lingual support:** I observed that a lot of non-english characters were included to the English tokenizer, including such as "ং": 1254, "অ": 1255 (devanagar), "ള": 1552,
      "വ": 1553 (Thai?), "ቀ": 1918, "ቁ": 1919 (Ethiopian),  "⌒": 2670, "⌘": 2671 (rare symbols), "╣": 2763, "╤": 2764 (graphics), "南": 3831, "単": 3832 (Chinese),       "😭": 8874,  "😰": 8875 (smileys).  "Llama 3 took 100K pre-trained tokens from [OpenAIs tiktoken library](https://github.com/openai/tiktoken/tree/main) and then trained it to add a further 28K tokens for better multi-lingual support." 
- **shared tokenizers**.  We wanted to have language specific decoders, but shared encoder.  Do we want to have a single tokenizer for all European languages or language-specific encoders and a language-specific encoder layer that interfaces the language to the shared encoder?
- **training data:**  I have downloaded 1 "shard" of Nemotron.  The first English shard has already 661 639 609 tokens and 9 372 643 (1,4%) space-separated token types -- far more than word forms in English.  My purpose now is to monitor the communication overhead, which already limits the amount of training that can be fully logged.  There is no point of using all the data or to produce a tokenizer based on it.  Is it the wise to use only the actual training (or training+validation) subsets to train the tokenizer?
- **available tokenizers:**   We did not find readily built tokenizer resources online.  Furthermore, Mammoth uses its own special tokens that might be missing in readily built tokenizers.   Thus, it is better to build my own tokenizers for Mammoth especially.
- **tokenizer training jobs:** Building tokenizers can take a good amount of time and it does not work in mobile online session.  So, I need to build a script that unzips the given text files and runs the 31 tokenizer learning tasks one after another on a CPU node and under nohup.  This should direct the progress bar etc.  to a null device.

Thus, I am trying to find answers to these questions and plan how to build 31 tokenizers and vocabularies for Nemotron models.  Only after this step, I can setup the config-config yaml file and run the experiments.

**Comment by jorgtied** [comment]
A few comments:
* I believe we moved away from sentence piece and use HF tokenizers now, don't we @chaowang0524 ?
* language-specific tokenizers: I would suggest 32K per language (not language pair)
* no normalization, please; maybe initial and trailing spaces can be removed but no lowercasing
* no pre-tokenization
* enabling backoff to bytes is always a good idea
* the data is messy and there will be non-standard characters in the data; just keep them
* training data: don't use the full shard for the first tests - use smaller subsets first (maybe 10 million lines per language?)
* shared tokenizer for the shared encoder: we could test both but in general a shared tokenizer across all languages supported by the encoder might work better; on the other hand, language-specific tokenizers and embeddings would lead to a model that can more easily be extended by yet another language; but anyway, try with a shared tokenizer first; The size of the vocab could be 128K (but that will depend a bit on the number of languages we support in our experiments)
* yes, build your own tokenizers; I have no idea why you want to run one after another when we have a cluster with plenty of CPU nodes available; simply submit all jobs to train the tokenizers all in parallel; one job per tokenizer; make sure that you use scratch folders or maybe even flash for fast I/O (the job should also do the unpacking and cleaning up)

**Comment by amikael** [comment]
[like]  Yli-Jyrä, Anssi reacted to your message:
________________________________

**Comment by chaowang0524** [comment]
> A few comments:
> 
> * I believe we moved away from sentence piece and use HF tokenizers now, don't we [@chaowang0524](https://github.com/chaowang0524) ?
> * language-specific tokenizers: I would suggest 32K per language (not language pair)
> * no normalization, please; maybe initial and trailing spaces can be removed but no lowercasing
> * no pre-tokenization
> * enabling backoff to bytes is always a good idea
> * the data is messy and there will be non-standard characters in the data; just keep them
> * training data: don't use the full shard for the first tests - use smaller subsets first (maybe 10 million lines per language?)
> * shared tokenizer for the shared encoder: we could test both but in general a shared tokenizer across all languages supported by the encoder might work better; on the other hand, language-specific tokenizers and embeddings would lead to a model that can more easily be extended by yet another language; but anyway, try with a shared tokenizer first; The size of the vocab could be 128K (but that will depend a bit on the number of languages we support in our experiments)
> * yes, build your own tokenizers; I have no idea why you want to run one after another when we have a cluster with plenty of CPU nodes available; simply submit all jobs to train the tokenizers all in parallel; one job per tokenizer; make sure that you use scratch folders or maybe even flash for fast I/O (the job should also do the unpacking and cleaning up)

I can answer the question 1-5:
1. Yes. We are using HF tokenizers in recent experiments. 
2. The size of the monolingual vocabs used in the recent experiments is exactly 32000 across languages now. 180000 was the number for the shared vocab we used to use before.
3. No normalization
4. HF tokenizer does not add word boundary mark automatically so currently I do pretokenize it with ("_") to indicate the word boundary like we did in spm. But I will look into it if pre-tokenization is not suggested.  Update: this might be a BART preference that inherented from the HF BART tokenizer. So it can be improved.
5. I'm not sure if HF tokenizer supports falling back to bytes. I haven't been working on this for a while, will definitely check later.

**Comment by jorgtied** [comment]
With pre-tokenization I meant any kind of fancy method for separating punctuation from words. I guess we do not want to do that, do we? sentence piece does not do that either, right?

**Comment by chaowang0524** [comment]
> With pre-tokenization I meant any kind of fancy method for separating punctuation from words. I guess we do not want to do that, do we? sentence piece does not do that either, right?

True, the pre-tokenizer I'm using only add "_" before the word boundary and it does not split punctuations from words.

**Comment by amikael** [comment]
Some comments.

1. ▁ refers to **sentencepiece model**, not Marian
2. To have bytefallback, we need **byte_fallback=True** to avoid <unk> on exotic code points.  `byte_fallback=True `works only in the Hugging Face tokenizers BPE model, not in SentencePiece tokenizers or SentencePieceProcessor models used by Marian, mT5, etc.
3. I agree that case matters, but we need to add NFKC normalization to eliminate variation with NFC vs NFD, Zero-width symboks.3. We need to set add_prefix_space=True for Metaspace to avoid generating separate tokens for the beginning of the sentence.
4. We do not want “Marian-style" where (`PAD==EOS==</s>` and no `<s>`.  Mammoth needs flexibility.
5. Even 10M lines of data will be too many for the extensive log files.  I will try 10000 although that is bad for translation quality.  And the experiment could use minimal, data tailored vocabs.

**Comment by jorgtied** [comment]
* @chaowang0524 - with word boundary - do you mean whitespace characters?
* about normalization: please, don't do any excessive normalization, even not NFKC or similar; any normalization will make the model weaker in handling noisy input that will have this kind of input and variation; the trend is to reduce normalization efforts and rather train models to become robust to this kind of data
* 10000 lines sounds super small but for some very limited testing this might work; but we need to also train more reasonable models now; would it be possible to have different levels of verbosity in logging? (depending on the purpose of the debugging / monitoring we want to do?)

**Comment by amikael** [comment]
DONE with tokenizers:
- modified Chaos BPE tokenization training script: args, byte_fallback, 32000.
- created 32000 token vocabs from 10000 lines (toy size data but good for monitoring comms) - small vocab with byte fallback. 
- fast to build language specific tokenizers: 4 mins each of languages
- 32000 tokens => all languages will need lack many full words.  
- can be extended/retrained later 

Model config questions (comparing prior Mammoth config files and Chao's):

- The following keys have meaningful changes:
```
model_dim: 768               # Transformer model dimension [model_dim: 512]
model_type: modernbert       #                             [model_type: text]
src_seq_length: 256          # [src_seq_length: 200]
tgt_seq_length: 256          # [tgt_seq_length: 200]
batch_size: 16                    # Per-GPU batch size (in sentences) [batch_size: 4096] !!!
valid_batch_size: 16              # Validation batch size             [valid_batch_size: 4096] !!!
batch_type: sents                 # Batch by sentences                [batch_type: tokens] !!!
normalization: tokens             # Normalize gradients by token count[normalization: tokens]
learning_rate: 0.001              # [learning_rate: 3.0] !!!
max_grad_norm: 1.0                # Gradient clipping per component [max_grad_norm: 0.0]
warmup_steps: 5000                # [warmup_steps: 10000]
decay_method: linear_warmup       # [decay_method: none]

```
- The following keys are no more in use in :

```
# [max_generator_batches: 2
# [src_subword_type: sentencepiece]
# [tgt_subword_type: sentencepiece]
# [skip_empty_level: silent]          # silently ignore empty lines in the data
# [src_vocab_size: 100000]
# [tgt_vocab_size: 100000]
# [param_init: 0.0]
# [param_init_glorot: true]
# [early_stopping: 5]
# [early_stopping_criteria: accuracy]
# [keep_checkpoint: -1]
# [save_all_gpus: false]
# [save_checkpoint_steps: 5000000]
# [encoder_type: transformer]
# [decoder_type: transformer]
# [transformer_ff: 2048]
```

It would be nice to document the changes and the reasons behind them.  I am not yet able to tell if some of these changes are due to the updates in the architecture, but there seem to be some that are just dropped.  The main new change is the key `x_transformers_opts`, but that already has a good explanation.
